language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -174,71 +263,89 @@ instantiateEvaluate con ty fun
             Just (IterApp headName spine) -> do
-                constAppResult <- applyStagedBuiltinName headName spine
+                constAppResult <- applyStagedBuiltinName arg headName spine
                 withVarEnv funVarEnv $ case constAppResult of
                     ConstAppSuccess res -> computeCek con res
                     ConstAppStuck       -> returnCek con term
 
-evaluateInCekM :: EvaluateConstApp (Either CekEvaluationException) -> CekM ConstAppResult
-evaluateInCekM a =
-    ReaderT $ \cekEnv ->
-        let eval means' term =
-                let cekEnv' = cekEnv & cekEnvMeans %~ mappend means'
-                    in runReaderT (computeCek [] term) cekEnv'
-            in runEvaluateT eval a
+-- | Reduce a saturated application of a builtin function in the empty context.
+computeInCekM :: EvaluateConstApp CekM ann -> CekM (ConstAppResult ann)
+computeInCekM = runEvaluateT eval where
+    eval means' = local (over cekEnvMeans $ mappend means') . computeCek [] . withMemory
 
 -- | Apply a 'StagedBuiltinName' to a list of 'Value's.
-applyStagedBuiltinName :: StagedBuiltinName -> [Plain Value] -> CekM ConstAppResult
-applyStagedBuiltinName (DynamicStagedBuiltinName name) args = do
+applyStagedBuiltinName
+    :: WithMemory Value
+    -> StagedBuiltinName
+    -> [WithMemory Value]
+    -> CekM (ConstAppResult ExMemory)
+applyStagedBuiltinName arg (DynamicStagedBuiltinName name) args = do
+    spendBoth arg 1 1
     DynamicBuiltinNameMeaning sch x <- lookupDynamicBuiltinName name
-    evaluateInCekM $ applyTypeSchemed sch x args
-applyStagedBuiltinName (StaticStagedBuiltinName  name) args =
-    evaluateInCekM $ applyBuiltinName name args
-
--- | Evaluate a term in an environment using the CEK machine.
-evaluateCekIn
-    :: CekEnv -> Plain Term -> Either CekEvaluationException (Plain Term)
-evaluateCekIn cekEnv = runCekM cekEnv . computeCek []
-
--- | Initialize an environment and evaluate a term in it using the CEK machine.
-evaluateCekInit
-    :: DynamicBuiltinNameMeanings -> Plain Term -> Either CekEvaluationException (Plain Term)
-evaluateCekInit means = evaluateCekIn (CekEnv means mempty)
+    fmap (fmap (const 1)) $ computeInCekM $ applyTypeSchemed
+        sch
+        x
+        (fmap void args)
+applyStagedBuiltinName arg (StaticStagedBuiltinName name) args = do
+    let (cpu, memory) = estimateStaticStagedCost name args
+    spendBoth arg cpu memory
+    fmap (fmap (const memory)) $ computeInCekM $ applyBuiltinName
+        name
+        (fmap void args)
+
+-- | Evaluate a term using the CEK machine and keep track of costing.
+runCek
+    :: DynamicBuiltinNameMeanings
+    -> CekBudgetMode
+    -> ExBudget
+    -> Plain Term
+    -> (Either CekEvaluationException (Plain Term), ExBudgetState)
+runCek means mode budget term =
+    runCekM (CekEnv means mempty mode)
+            (ExBudgetState mempty (budget <> ExBudget 0 (modeModifier $ termAnn memTerm)))
```

Comment:
It took me a few minutes to work out what's going on here. Comment please!

Reply:
Should be quite a bit simpler now. If not, unresolve please.